keps/sig-auth/3766-referential-authorization/README.md

```diff
@@ -413,80 +422,115 @@ rejected".
+// Subject refers to the subject that is a consumer of the referenced
+// pattern(s).
+type ClusterReferenceConsumer struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // Subject refers to the subject that is a consumer of the referenced
+    // pattern(s).
+    Subject Subject       `json:"subject"`
+
+    // From refers to the group and resource that these references originate from.
+    From    GroupResource `json:"from"`
+
+    // To refers to the group and resource that these references target.
+    To      GroupResource `json:"to"`
+
+    // For refers to the purpose of this reference. (Cluster)ReferenceGrants
+    // matching the From, To, and For of this resource will be authorized for
+    // the Subject of this resource.
+    For     string        `json:"for"`
+}
 
-	// Spec defines the desired state of ReferenceGrant.
-	Spec ReferenceGrantSpec `json:"spec,omitempty"`
+// Subject is a copy of RBAC Subject that excludes APIGroup.
+type Subject struct {
+    // Kind of object being referenced. Values defined by this API group are
+    // "User", "Group", and "ServiceAccount". If the Authorizer does not
+    // recognized the kind value, the Authorizer should report an error.
+    Kind string `json:"kind"`
+    // Name of the object being referenced.
+    Name string `json:"name"`
+    // Namespace of the referenced object.  If the object kind is non-namespace,
+    // such as "User" or "Group", and this value is not empty the Authorizer
+    // should report an error. +optional
+    Namespace string `json:"namespace,omitempty"`
+}
+```
 
-	// Note that `Status` sub-resource has been excluded at the
-	// moment as it was difficult to work out the design.
-	// A `Status` sub-resource may be added in the future.
+#### ClusterReferenceGrant
+
+```golang
+// ClusterReferenceGrant authorizes same-namespace references matching the
+// specified pattern.
+type ClusterReferenceGrant struct {
+    metav1.TypeMeta   `json:",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    // From refers to the group, resource, and pthat that these references
+    // originate from.
+    From    GroupVersionResourcePath `json:"from"`
+
+    // To refers to the group and resource that these references target.
+    To      GroupResource `json:"to"`
+
+    // For refers to the purpose of this reference. Subjects of
+    // ClusterReferenceConsumers will be authorized to follow references matching
+    // the From, To, and For of this resource.
+    For     string        `json:"for"`
 }
 
-// +kubebuilder:object:root=true
-// ReferenceGrantList contains a list of ReferenceGrant.
-type ReferenceGrantList struct {
-	metav1.TypeMeta `json:",inline"`
-	metav1.ListMeta `json:"metadata,omitempty"`
-	Items           []ReferenceGrant `json:"items"`
+type GroupVersionResourcePath struct {
+    Group    string `json:"group"`
+    Resource string `json:"resource"`
+    Version  string `json:"version"`
```

Comment:
Why do we specify version here? Just for being super safe on fully-qualifying everything, or is there a risk that `/apis/foo.com/v1/my-resource` and `/apis/foo.com/v2/my-resource` would serve completely different things, such that references would break?

In other places I see only GroupResource, why is `ClusterReferenceGrant.from` different?

Reply:
The controller/authorizer has to watch the full contents of these resources to follow the Path to retrieve the names of each target resource we would like to granularly authorize.

Each Path is going to be specific to the structure of that versioned API.
The references would break between versions like you suggest.

We need to support multiple versions here.

A ClusterReferenceGrant could support multiple GroupVersionResourcePaths in a list.
The alternative is that multiple versions would have to be supported with multiple resources.

Reply:
+1 to what @stealthybox said, JSONPath could be interpreted differently depending on the API Version used.

Reply:
True, thanks for pointing out that the path (of course) is version-specific, makes sense like this then.

Then the bigger question is that shall we support a list of GroupVersionResourcePaths or not? I think maybe yes, although it increases the complexity a little, as it's no fun having to manage multiple API objects on upgrades.